miniapps/tools/nodal_transfer.cpp

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+//
+// This miniapp projects a set of functions, defined by a set of points and
+// values evaluated at these points, onto a grid function. The main assumption
+// is that the nodes of the target grid function coincide with some of the
+// evaluation points. The miniapp has two modes. The first mode generates the
+// data and writes it to a set of data files. The data can be generated on any
+// number of processes. The second mode reads the saved data and projects it
+// on the newly defined grid function. The number of processes in the second
+// mode can differ from the number of processes used for the set generation.
+//
+// Sample runs:
+// Generate second order mesh on 4 processes
+// mpirun -np 4 ./nodal_transfer -rs 2 -rp 1 -gd 1 -o 2
+// Read the generated data and map it to a grid function
+// defined on two processes
+// mpirun -np 2 ./nodal_transfer -rs 2 -rp 0 -gd 0 -snp 4 -o 2
+//
+// Generate first order grid function on 8 processes
+// mpirun -np 8 ./nodal_transfer -rs 2 -rp 2 -gd 1 -o 1 -m ../../data/star.mesh
+// Read the generated data on 4 processes and coarser mesh
+// mpirun -np 4 ./nodal_transfer -rs 2 -rp 0 -gd 0
+//                            -snp 8 -o 1 -m ../../data/star.mesh
+//
+
+#include <mfem.hpp>
+#include <fstream>
+#include <iostream>
+#include "../common/mfem-common.hpp"
+#include <cmath>
+
+using namespace mfem;
+
+class TestCoeff:public Coefficient
+{
+public:
+   TestCoeff()
+   {
+
+   }
+
+   virtual
+   double Eval(ElementTransformation &T,
+               const IntegrationPoint &ip)
+   {
+      if (T.GetSpaceDim()==3)
+      {
+         double x[3];
+         Vector transip(x, 3);
+         T.Transform(ip, transip);
+         return std::sin(x[0])*std::cos(x[1]) +
+                std::sin(x[1])*std::cos(x[2]) +
+                std::sin(x[2])*std::cos(x[0]);
+      }
+      else
+      {
+         double x[2];
+         Vector transip(x, 2);
+         T.Transform(ip, transip);
+         return std::sin(x[0])*std::cos(x[1]) +
+                std::sin(x[1])*std::cos(x[0]);
+      }
+   }
```

Comment:
Can we write this as a dimension independent code instead of branches? This way we can also cover the 1D case for test coverage and possibly future extension for 4D.

Reply:
I will add 1D branch. I am not sure if the mfem data structures support points with dimension larger than 3.